Description: EZC: clear EG(exception) on return from zend_wrap_func()
 Otherwise, if the exception is caught and then control is returned into
 a zend_call_function() call which wrapped the throwing function, then
 zend_call_function() will erroneously rethrow the exception, even though
 it has already been handled.
 See  
Author: Tim Starling <tstarling@wikimedia.org>
Last-Update: 2014-09-25
Origin: backport, https://reviews.facebook.net/D23883
Bug: https://bugzilla.wikimedia.org/show_bug.cgi?id=71174
Reviewed-by: Giuseppe Lavagetto <glavagetto@wikimedia.org>

Index hhvm/hphp/runtime/ext_zend_compat/hhvm/zend-wrap-func.cpp
===================================================================
--- hhvm.orig/hphp/runtime/ext_zend_compat/hhvm/zend-wrap-func.cpp
+++ hhvm/hphp/runtime/ext_zend_compat/hhvm/zend-wrap-func.cpp
@@ -20,6 +20,7 @@
 #include "hphp/runtime/ext_zend_compat/php-src/Zend/zend.h"
 #include "hphp/runtime/ext_zend_compat/php-src/TSRM/TSRM.h"
 #include "hphp/runtime/ext_zend_compat/php-src/Zend/zend_exceptions.h"
+#include "hphp/runtime/ext_zend_compat/php-src/Zend/zend_globals.h"
 #include "hphp/runtime/vm/jit/translator-inline.h"
 
 namespace HPHP {
@@ -42,6 +43,14 @@ void zBoxAndProxy(TypedValue* arg) {
   }
 }
 
+static void zend_wrap_func_cleanup() {
+    ZendExecutionStack::popHHVMStack();
+    if (EG(exception)) {
+      zval_ptr_dtor(&EG(exception));
+      EG(exception) = NULL;
+    }
+}
+
 TypedValue* zend_wrap_func(ActRec* ar) {
   // Sync the translator state.
   // We need to do this before a native function calls into a C library
@@ -86,10 +95,10 @@ TypedValue* zend_wrap_func(ActRec* ar) {
       TSRMLS_CC
     );
   } catch (...) {
-    ZendExecutionStack::popHHVMStack();
+    zend_wrap_func_cleanup();
     throw;
   }
-  ZendExecutionStack::popHHVMStack();
+  zend_wrap_func_cleanup();
 
   // If an exception was caught, rethrow it
   ZendExceptionStore& exceptionStore = ZendExceptionStore::getInstance();

Index hhvm/hphp/test/slow/ext_ezc_test/caught-exception-recursive.expectf
===================================================================
--- /dev/null
+++ hhvm/hphp/test/slow/ext_ezc_test/caught-exception-recursive.expectf
@@ -0,0 +1 @@
+bool(false)
Index hhvm/hphp/test/slow/ext_ezc_test/caught-exception-recursive.php
===================================================================
--- /dev/null
+++ hhvm/hphp/test/slow/ext_ezc_test/caught-exception-recursive.php
@@ -0,0 +1,7 @@
+<?php
+var_dump(ezc_try_call(function () {
+  try {
+    ezc_throw('Exception');
+  } catch ( Exception $e ) {}
+  return false;
+}));
